File: src/main/java/frc/robot/io/TalonFXMotorIO.java
package frc.robot.io;

import static frc.robot.util.constants.IntakeConstants.INTAKE_ARM_DEPLOYED_POSITION;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.FeedbackConfigs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.DutyCycleOut;
import com.ctre.phoenix6.controls.Follower;
import com.ctre.phoenix6.controls.MotionMagicExpoTorqueCurrentFOC;
import com.ctre.phoenix6.controls.MotionMagicVelocityTorqueCurrentFOC;
import com.ctre.phoenix6.controls.PositionVoltage;
import com.ctre.phoenix6.controls.TorqueCurrentFOC;
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC;
import com.ctre.phoenix6.controls.VoltageOut;
import com.ctre.phoenix6.hardware.CANcoder;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.FeedbackSensorSourceValue;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Temperature;
import edu.wpi.first.units.measure.Voltage;
import java.util.Optional;

/**
 * CTRE Talon FX implementation of {@link MotorIO}.
 *
 * <p>Velocity commands are accepted in RPM and converted to rotations per second for Phoenix.
 * Position commands use mechanism rotations. Status signals are registered with {@link
 * SignalRegistry} for batched refresh before subsystem inputs are read.
 */
public class TalonFXMotorIO implements TorqueCurrentMotorIO {
  private static final double CONTROL_SIGNAL_FREQUENCY_HZ = 100.0;
  private static final double TEMPERATURE_SIGNAL_FREQUENCY_HZ = 0.25;

  protected final TalonFX motor;
  protected final int canID;
  protected final TalonFXConfiguration config;
  protected final String motorName;
  protected CANcoder canCoder;

  protected final VelocityTorqueCurrentFOC velocityRequest;
  protected final MotionMagicVelocityTorqueCurrentFOC mmVelocityRequest;
  protected final MotionMagicExpoTorqueCurrentFOC mmPositionRequest;
  protected final PositionVoltage positionVoltageRequest;
  protected final DutyCycleOut dutyCycleRequest;
  protected final VoltageOut voltageRequest;
  protected final TorqueCurrentFOC torqueCurrentRequest;

  /** Cached status signals refreshed in one batch call to reduce CAN bus load. */
  protected final StatusSignal<Voltage> motorVoltageSignal;

  protected final StatusSignal<Current> statorCurrentSignal;
  protected final StatusSignal<Temperature> deviceTempSignal;
  protected final StatusSignal<AngularVelocity> velocitySignal;
  protected final StatusSignal<Angle> positionSignal;

  /** Signals needed for follower monitoring and torque-current telemetry. */
  protected final StatusSignal<Double> dutyCycleSignal;

  protected final StatusSignal<Current> torqueCurrentSignal;

  /** Standard motor (no follower, no CANcoder). */
  public TalonFXMotorIO(int canID, TalonFXConfiguration config, String motorName) {
    this(canID, config, motorName, Optional.empty(), Optional.empty());
  }

  /** Follower motor. */
  public TalonFXMotorIO(
      int canID, TalonFXConfiguration config, String motorName, Follower followerConfig) {
    this(canID, config, motorName, Optional.empty(), Optional.of(followerConfig));
  }

  /** Motor with external CANcoder feedback. */
  public TalonFXMotorIO(
      int canID,
      TalonFXConfiguration config,
      String motorName,
      CANcoderConfiguration canCoderConfig) {
    this(canID, config, motorName, Optional.of(canCoderConfig), Optional.empty());
  }

  /** Full constructor. */
  public TalonFXMotorIO(
      int canID,
      TalonFXConfiguration config,
      String motorName,
      Optional<CANcoderConfiguration> canCoderConfig,
      Optional<Follower> followerConfig) {
    this.canID = canID;
    this.config = config;
    this.motorName = motorName;

    motor = new TalonFX(canID);
    motor.clearStickyFaults();

    velocityRequest = new VelocityTorqueCurrentFOC(0);
    mmVelocityRequest = new MotionMagicVelocityTorqueCurrentFOC(0);
    mmPositionRequest = new MotionMagicExpoTorqueCurrentFOC(0);
    positionVoltageRequest = new PositionVoltage(INTAKE_ARM_DEPLOYED_POSITION).withEnableFOC(true);
    dutyCycleRequest = new DutyCycleOut(0).withEnableFOC(true);
    voltageRequest = new VoltageOut(0).withEnableFOC(true);
    torqueCurrentRequest = new TorqueCurrentFOC(0);

    motor.optimizeBusUtilization(0);

    motorVoltageSignal = motor.getMotorVoltage();
    statorCurrentSignal = motor.getStatorCurrent();
    deviceTempSignal = motor.getDeviceTemp();
    velocitySignal = motor.getVelocity();
    positionSignal = motor.getPosition();

    dutyCycleSignal = motor.getDutyCycle();
    torqueCurrentSignal = motor.getTorqueCurrent();

    configureStatusSignalFrequencies();
    SignalRegistry.getInstance().registerMotorIO(this);
    configureRemoteCancoder(canCoderConfig);

    if (canCoderConfig.isEmpty()) {
      motor.getConfigurator().apply(config);
    }

    followerConfig.ifPresent(motor::setControl);
  }

  private void configureStatusSignalFrequencies() {
    BaseStatusSignal.setUpdateFrequencyForAll(
        CONTROL_SIGNAL_FREQUENCY_HZ,
        dutyCycleSignal,
        motorVoltageSignal,
        torqueCurrentSignal,
        statorCurrentSignal,
        velocitySignal,
        positionSignal);

    deviceTempSignal.setUpdateFrequency(TEMPERATURE_SIGNAL_FREQUENCY_HZ);
  }

  private void configureRemoteCancoder(Optional<CANcoderConfiguration> canCoderConfig) {
    canCoderConfig.ifPresent(
        ccCfg -> {
          canCoder = new CANcoder(canID);
          canCoder.getConfigurator().apply(ccCfg);
          TalonFXConfiguration updatedConfig =
              config.withFeedback(
                  new FeedbackConfigs()
                      .withFeedbackRemoteSensorID(canCoder.getDeviceID())
                      .withFeedbackSensorSource(FeedbackSensorSourceValue.FusedCANcoder));
          motor.getConfigurator().apply(updatedConfig);

          canCoder.optimizeBusUtilization(0);

          BaseStatusSignal.setUpdateFrequencyForAll(
              CONTROL_SIGNAL_FREQUENCY_HZ,
              canCoder.getPosition(),
              canCoder.getVelocity(),
              canCoder.getAbsolutePosition());
        });
  }

  @Override
  public String getMotorName() {
    return motorName;
  }

  @Override
  public void setMotorVoltage(Voltage voltage) {
    motor.setControl(voltageRequest.withOutput(voltage));
  }

  @Override
  public void setMotorPercentage(double percentage) {
    motor.setControl(dutyCycleRequest.withOutput(percentage));
  }

  /** Converts RPM to rotations per second for Phoenix velocity closed loop. */
  @Override
  public void setMotorRPM(double rpm) {
    motor.setControl(velocityRequest.withVelocity(rpm / 60.0));
  }

  @Override
  public void setMotorPosition(double setpoint) {
    motor.setControl(positionVoltageRequest.withPosition(setpoint));
  }

  @Override
  public void setMotorPosition(double setpoint, int slotID) {
    motor.setControl(positionVoltageRequest.withPosition(setpoint).withSlot(slotID));
  }

  @Override
  public void setMotorTorqueCurrent(Current torqueCurrent) {
    motor.setControl(torqueCurrentRequest.withOutput(torqueCurrent));
  }

  @Override
  public BaseStatusSignal[] getStatusSignals() {
    return new BaseStatusSignal[] {
      motorVoltageSignal,
      statorCurrentSignal,
      deviceTempSignal,
      velocitySignal,
      positionSignal,
      dutyCycleSignal,
      torqueCurrentSignal
    };
  }

  @Override
  public void updateInputs(MotorIOInputs inputs) {
    inputs.setMotorConnected(
        BaseStatusSignal.isAllGood(
            motorVoltageSignal, statorCurrentSignal, velocitySignal, positionSignal));
    inputs.setMotorVoltage(motorVoltageSignal.getValueAsDouble());
    inputs.setMotorCurrent(statorCurrentSignal.getValueAsDouble());
    inputs.setMotorTemperature(deviceTempSignal.getValueAsDouble());
    inputs.setMotorVelocity(velocitySignal.getValueAsDouble());
    inputs.setMotorPosition(positionSignal.getValueAsDouble());
  }

  /** Returns the backing TalonFX for simulation wrappers and tuning helpers. */
  public TalonFX getMotor() {
    return motor;
  }
}

File: src/main/cpp/io/TalonFXMotorIO.cpp
#include "io/TalonFXMotorIO.h"
#include "Constants.h"

#include <ctre/phoenix6/BaseStatusSignal.hpp>

using namespace ctre::phoenix6;

TalonFXMotorIO::TalonFXMotorIO(int canID,
                                 configs::TalonFXConfiguration config,
                                 std::string motorName)
    : TalonFXMotorIO(canID, std::move(config), std::move(motorName),
                     std::nullopt, std::nullopt) {}

TalonFXMotorIO::TalonFXMotorIO(int canID,
                                 configs::TalonFXConfiguration config,
                                 std::string motorName,
                                 controls::Follower followerConfig)
    : TalonFXMotorIO(canID, std::move(config), std::move(motorName),
                     std::nullopt, std::make_optional(std::move(followerConfig))) {}

TalonFXMotorIO::TalonFXMotorIO(int canID,
                                 configs::TalonFXConfiguration config,
                                 std::string motorName,
                                 configs::CANcoderConfiguration canCoderConfig)
    : TalonFXMotorIO(canID, std::move(config), std::move(motorName),
                     std::make_optional(std::move(canCoderConfig)), std::nullopt) {}

TalonFXMotorIO::TalonFXMotorIO(
    int canID,
    configs::TalonFXConfiguration config,
    std::string motorName,
    std::optional<configs::CANcoderConfiguration> canCoderConfig,
    std::optional<controls::Follower> followerConfig)
    : m_motor{canID},
      m_canID{canID},
      m_config{std::move(config)},
      m_motorName{std::move(motorName)},
      m_motorVoltageSignal{m_motor.GetMotorVoltage()},
      m_statorCurrentSignal{m_motor.GetStatorCurrent()},
      m_deviceTempSignal{m_motor.GetDeviceTemp()},
      m_velocitySignal{m_motor.GetVelocity()},
      m_positionSignal{m_motor.GetPosition()},
      m_dutyCycleSignal{m_motor.GetDutyCycle()},
      m_torqueCurrentSignal{m_motor.GetTorqueCurrent()} {
  m_motor.ClearStickyFaults();

  // Disable all unused signals first
  m_motor.OptimizeBusUtilization(0_Hz);

  // Re-enable update for signals that we are using
  BaseStatusSignal::SetUpdateFrequencyForAll(
      100_Hz,
      m_dutyCycleSignal,
      m_motorVoltageSignal,
      m_torqueCurrentSignal,
      m_statorCurrentSignal,
      m_velocitySignal,
      m_positionSignal);

  m_deviceTempSignal.SetUpdateFrequency(0.25_Hz);

  // Register the motor to the signal registry
  SignalRegistry::GetInstance().RegisterMotorIO(*this);

  // Apply CANcoder config if present
  if (canCoderConfig.has_value()) {
    m_canCoder = std::make_unique<hardware::CANcoder>(canID);
    m_canCoder->GetConfigurator().Apply(canCoderConfig.value());

    configs::TalonFXConfiguration updatedConfig = m_config;
    updatedConfig.Feedback.FeedbackRemoteSensorID = m_canCoder->GetDeviceID();
    updatedConfig.Feedback.FeedbackSensorSource =
        signals::FeedbackSensorSourceValue::FusedCANcoder;
    m_motor.GetConfigurator().Apply(updatedConfig);

    m_canCoder->OptimizeBusUtilization(0_Hz);

    BaseStatusSignal::SetUpdateFrequencyForAll(
        100_Hz,
        m_canCoder->GetPosition(),
        m_canCoder->GetVelocity(),
        m_canCoder->GetAbsolutePosition());
  } else {
    m_motor.GetConfigurator().Apply(m_config);
  }

  // Apply follower config if present
  if (followerConfig.has_value()) {
    m_motor.SetControl(followerConfig.value());
  }
}

std::string TalonFXMotorIO::GetMotorName() const {
  return m_motorName;
}

void TalonFXMotorIO::SetMotorVoltage(double voltage) {
  m_motor.SetVoltage(units::voltage::volt_t{voltage});
}

void TalonFXMotorIO::SetMotorPercentage(double percentage) {
  m_motor.Set(percentage);
}

void TalonFXMotorIO::SetMotorRPM(double rpm) {
  m_motor.SetControl(m_velocityRequest.WithVelocity(
      units::angular_velocity::turns_per_second_t{rpm / 60.0}));
}

void TalonFXMotorIO::SetMotorPosition(double setpoint) {
  m_motor.SetControl(m_positionVoltageRequest.WithPosition(
      units::angle::turn_t{setpoint}));
}

void TalonFXMotorIO::SetMotorPosition(double setpoint, int slotID) {
  m_motor.SetControl(m_positionVoltageRequest
      .WithPosition(units::angle::turn_t{setpoint})
      .WithSlot(slotID));
}

std::vector<BaseStatusSignal*> TalonFXMotorIO::GetStatusSignals() {
  return {
    &m_motorVoltageSignal,
    &m_statorCurrentSignal,
    &m_deviceTempSignal,
    &m_velocitySignal,
    &m_positionSignal,
    &m_dutyCycleSignal,
    &m_torqueCurrentSignal
  };
}

void TalonFXMotorIO::UpdateInputs(MotorIOInputs& inputs) {
  inputs.motorConnected = BaseStatusSignal::IsAllGood(
      m_motorVoltageSignal, m_statorCurrentSignal,
      m_velocitySignal, m_positionSignal);
  inputs.motorVoltage = m_motorVoltageSignal.GetValueAsDouble();
  inputs.motorCurrent = m_statorCurrentSignal.GetValueAsDouble();
  inputs.motorTemperature = m_deviceTempSignal.GetValueAsDouble();
  inputs.motorVelocity = m_velocitySignal.GetValueAsDouble();
  inputs.motorPosition = m_positionSignal.GetValueAsDouble();
}
